### ISSUE TITLE
Prevent uninstallation of Microsoft Edge stable

### ISSUE THREAD

**Issue by pratyakshm** [microsoft/winget-cli#1286]
# Update
I propose an updated and better implementation:
If the user tries to uninstall Edge, warn them that Edge is a dependency and uninstalling it might cause system issues. Allow them to override the warning and proceed with uninstallation.

# Description of the new feature/enhancement
I request that WinGet must block uninstallation of Microsoft Edge stable since the browser is basically tightly integrated to the OS and uninstalling it might cause OS instability including but not limited to issues like failed Windows Updates and broken features. 

# Proposed technical implementation details (optional)
I do not have any code snippets to share at this moment. But a technical thing I suggest is printing an error message when ``winget uninstall Microsoft.Edge`` (or any similar command to uninstall the stable version of Microsoft Edge) is used. Maybe something simple like "WinGet does not support uninstallation of this app".

**Comment by Masamune3210** [comment]
I say leave it. if you are smart enough to know how to manually uninstall it using a shell, you are probably smart enough to figure out how to re-register it too

**Comment by Trenly** [comment]
> I say leave it. if you are smart enough to know how to manually uninstall it using a shell, you are probably smart enough to figure out how to re-register it too

I tend to agree that if users want to be able to remove it, they should be able to. However, I don't think it should be without warning. I could easily see a young tech enthusiast who always uses Google Chrome or Firefox thinking they would be smart and uninstall Microsoft Edge, not realizing how many windows features depend on edge. 

I would propose that in this specific case, the user be warned that Edge is a dependency for Windows features, and removing it may cause system instability or remove access to some features, requiring a confirmation before proceeding with the uninstall

**Comment by Masamune3210** [comment]
That I will agree with being a good idea, not everyone is ready to learn by burning down their boat so to speak

**Comment by zhangzp9970** [comment]
I have just uninstalled MS edge using winget, I'm now finding a way to reinstall it as a system app

**Comment by zhangzp9970** [comment]
Just download one from MS store and everything will be back.

**Comment by lackovic** [comment]
Uninstallation of Microsoft Edge fails with exit code 93.

```
> winget uninstall -e --id Microsoft.Edge
Found Microsoft Edge [Microsoft.Edge]
Starting package uninstall...
Uninstall failed with exit code: 93

> systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.19045 N/A Build 19045
```

**Comment by pratyakshm** [comment]
I agree with @Trenly. 

I've changed my stance since I originally posted the issue. The user must have as much control over their device as possible. 

I propose an updated and better implementation:
If the user tries to uninstall Edge, warn them that Edge is a dependency and uninstalling it might cause system issues. Allow them to override the warning and proceed with uninstallation.

**Comment by luckybear97** [comment]
EDGE DEV >>>>>>>>>>>>>>>> EDGE

crazy that we need two browsers just cause it's the default APP!

**Comment by cpprust** [comment]
Despite setting Firefox as my default browser, Microsoft Edge still opens when I press <F1> on my desktop.
To try and fix this issue, I have decided to remove Microsoft Edge.
If I need to reinstall it, I can still use the command `winget install microsoft.edge`. (So I think it should be safe?)
Unfortunately, I am not able to remove it via `winget uninstall microsoft.edge`.

**Comment by pratyakshm** [comment]
> Despite setting Firefox as my default browser, Microsoft Edge still opens when I press <F1> on my desktop.
> To try and fix this issue, I have decided to remove Microsoft Edge.
> If I need to reinstall it, I can still use the command `winget install microsoft.edge`. (So I think it should be safe?)
> Unfortunately, I am not able to remove it via `winget uninstall microsoft.edge`.

@cpprust does it throw any error if you try to run:
```powershell
winget uninstall Microsoft.Edge
```
?

**Comment by luckybear97** [comment]
> > Despite setting Firefox as my default browser, Microsoft Edge still opens when I press  on my desktop.
> > To try and fix this issue, I have decided to remove Microsoft Edge.
> > If I need to reinstall it, I can still use the command `winget install microsoft.edge`. (So I think it should be safe?)
> > Unfortunately, I am not able to remove it via `winget uninstall microsoft.edge`.
> 
> @cpprust does it throw any error if you try to run:
> 
> ```powershell
> winget uninstall Microsoft.Edge
> ```
> 
> ?

Don't work, they patched it. Technically you still can remove the files manually, but yeah, I don't want to take risk

**Comment by cpprust** [comment]
@pratyakshm

```
> winget remove microsoft.edge
Found Microsoft Edge [Microsoft.Edge]
Starting package uninstall...
Uninstall failed with exit code: 93
```

**Comment by pratyakshm** [comment]
Interesting.

@denelon Can you confirm if WinGet, by behavior, blocks Edge Stable uninstallation?

**Comment by pratyakshm** [comment]
Understood. This is on Windows client side then.

Thanks for the information!

**Comment by Taky0nDon** [comment]
> Interesting.
> 
> @denelon Can you confirm if WinGet, by behavior, blocks Edge Stable uninstallation?

I just uninstalled Edge Stable earlier today, found with issue while trying to figure out what error 93 was.

**Comment by 3xploiton3** [comment]
winget uninstall Microsoft.EdgeWebView2Runtime
Found Microsoft Edge WebView2 Runtime [Microsoft.EdgeWebView2Runtime]
Starting package uninstall...
Uninstall failed with exit code: 93

winget uninstall Microsoft.Edge
Found Microsoft Edge [Microsoft.Edge]
Starting package uninstall...
Uninstall failed with exit code: 93

![image](https://github.com/microsoft/winget-cli/assets/19517680/233705e7-8a6e-4654-93aa-e59eb94258e1)

**Comment by Kazlehoff** [comment]
Googling Uninstall failed with exit code: 93 brought me here. guess why. 

its crap like this that makes me wish linux was better for a daily driver for me. :(

**Comment by metamec** [comment]
Yet another Microsoft application making winget a pain to use.

**Comment by jimmy-zhening-luo** [comment]
I'm not going to lie... former Microsoft employee here, and I got here because I was trying to uninstall Microsoft.Edge and wanted to know what error 93 was ;)

(That said, I already had a suspicion that uninstalling Microsoft.Edge would break Windows but hoped beyond hope that Edgium, WebPlat, WinINet, WinHTTP, and friends finally had enough separation from the rest of Shell/NT after all these years.)

For future readers who find this thread, trust me, you do **not** want to uninstall Microsoft.Edge or the WebView Runtime, even though that's what I was trying to do. (Frankly as someone who also worked on AAD(EntraId?TFS?VSO?VSTS?ADO?AzureBoard?)/OneAuth at one point, I have no idea why I thought it'd be a good idea to try to uninstall the webivew lolol.)

**Forget about warnings or configurability, I think period, Microsoft Edge should not be uninstallable.**

### Root Cause

Why is that? Does Microsoft hate you? Are we the Evil Empire ™? Maybe so re: the latter, definitely not re: the former (I think most of the folks I worked with did genuinely love our users). No, rather, you can thank your local congressperson and Department of Justice official for their brilliant performance in [United States of America v. Microsoft Corporation, 253 F.3d 34](https://en.wikipedia.org/wiki/United_States_v._Microsoft_Corp.) regarding the "bundling" of Internet Explorer with Microsoft, and the public/private API hell that emerged from it. You can choose to believe however that everyone at Microsoft is Bad At Building Operating Systems, as a proud former Windows eng I'm not even sure I'd argue with on on that.

**Comment by abowen** [comment]
You need to use WinUtil, go to Tweaks section, and select Remove Microsoft Edge down the bottom left, then run. 

It's bunch of scripts to successfully uninstall Edge

https://github.com/ChrisTitusTech/winutil

I would have been fine leaving it on my PC but I set Firefox as my default browser, and Microsoft apps choose to ignore that
So now I'm removing it on principle from every PC I support

**Comment by jimmy-zhening-luo** [comment]
> > **I think period, Microsoft Edge should not be uninstallable.**
> 
> As per my earlier message, [Microsoft stated this year in February](https://blogs.windows.com/windows-insider/2023/11/16/previewing-changes-in-windows-to-comply-with-the-digital-markets-act-in-the-european-economic-area/) "_for compliance with the Digital Markets Act (DMA) [...] we’ve added the ability to uninstall Microsoft Edge, in the EEA_".

Ah I didn't know that. I'm surprised that's a DMA compliance item. Thanks for correcting me on this though, that's an important exception -- I retract my opinion that Microsoft Edge should not be uninstallable.

**Comment by leo60228** [comment]
Worth noting: if both the MSI and Store versions of Edge are installed, either can be uninstalled. (To make `winget upgrade --id Microsoft.Edge` work on a fresh installation, you actually need to `winget install --id Microsoft.Edge --force` then uninstall the stock Edge...)
